File: Coin Change Problem/coinChange.java
import java.io.*;
import java.lang.*;
import java.util.*;


class Main{
	//This function finds the minimum number of coins required to make an amount else returns -1.
	public static int coinChange(int[] coins,int amount){
		Arrays.sort(coins);
		int[] dp = new int[amount+1];
		Arrays.fill(dp,amount+1);
		dp[0] = 0;
		for (int i=1;i<=amount;i++){
			for (int j=0;j<coins.length;j++){
				if (coins[j] <= i){
					dp[i] = Math.min(dp[i], 1 + dp[i - coins[j]]);
				}else{
					break;
				}
			}
		}
		return dp[amount] == amount + 1 ? -1 : dp[amount];
	}
	//Main Function.
	public static void main(String[] args){
		int[] coins = {1,2,5};
		int amount = 11;
		int ans = coinChange(coins,amount);
		System.out.println(ans);
	}
}

File: Coin Change Problem/coinChange.py
# Calculate minimum number of coins to get the required amount
def coinChange(coins, amount):
    MAX = float('inf')
    minCoinCount = [0] + [MAX] * (amount + 1)

    for currentAmount in range(0, amount + 1):
        for coin in coins:
            if coin <= currentAmount:
                minCoinCount[currentAmount] = min(minCoinCount[currentAmount], 1 + minCoinCount[currentAmount - coin])

    if minCoinCount[amount] == MAX:
        return -1

    return minCoinCount[amount]


if __name__ == "__main__":
    array = []

    # Inputs
    coins = [1, 2, 5]
    amount = 11

    print(coinChange(coins, amount))
